mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityHelperSchemaValidationMethodGenerator.java

```diff
@@ -0,0 +1,309 @@
+
+import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata;
+import com.datastax.oss.driver.api.core.type.UserDefinedType;
+import com.datastax.oss.driver.api.core.type.reflect.GenericType;
+import com.datastax.oss.driver.api.mapper.annotations.SchemaHint;
+import com.datastax.oss.driver.internal.core.metadata.schema.DefaultTableMetadata;
+import com.datastax.oss.driver.internal.mapper.processor.MethodGenerator;
+import com.datastax.oss.driver.internal.mapper.processor.dao.LoggingGenerator;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.TypeName;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.TypeElement;
+
+public class EntityHelperSchemaValidationMethodGenerator implements MethodGenerator {
+
+  private final EntityDefinition entityDefinition;
+  private TypeElement entityTypeElement;
+  private LoggingGenerator loggingGenerator;
+  private EntityHelperGenerator entityHelperGenerator;
+
+  public EntityHelperSchemaValidationMethodGenerator(
+      EntityDefinition entityDefinition,
+      TypeElement entityTypeElement,
+      LoggingGenerator loggingGenerator,
+      EntityHelperGenerator entityHelperGenerator) {
+    this.entityDefinition = entityDefinition;
+    this.entityTypeElement = entityTypeElement;
+    this.loggingGenerator = loggingGenerator;
+    this.entityHelperGenerator = entityHelperGenerator;
+  }
+
+  @Override
+  public Optional<MethodSpec> generate() {
+    MethodSpec.Builder methodBuilder =
+        MethodSpec.methodBuilder("validateEntityFields")
+            .addAnnotation(Override.class)
+            .addModifiers(Modifier.PUBLIC)
+            .returns(TypeName.VOID);
+
+    // get keyspaceId from context, and if not present fallback to keyspace set on session
+    methodBuilder.addStatement(
+        "$1T keyspaceId = context.getKeyspaceId() != null ? context.getKeyspaceId() : context.getSession().getKeyspace().orElse(null)",
+        CqlIdentifier.class);
+
+    methodBuilder.addStatement("String entityClassName = $S", entityDefinition.getClassName());
+
+    // Handle case where keyspaceId = null.
+    // In such case we cannot infer and validate schema for table or udt
+    methodBuilder.beginControlFlow(
+        "if (!context.getSession().getMetadata().getKeyspace(keyspaceId).isPresent())");
+    loggingGenerator.warn(
+        methodBuilder,
+        "[{}] Unable to validate table: {} for the entity class: {} because keyspace: {} is not present",
+        CodeBlock.of("context.getSession().getName()"),
+        CodeBlock.of("tableId"),
+        CodeBlock.of("entityClassName"),
+        CodeBlock.of("keyspaceId"));
```

Comment:
The error looks like this:
`Unable to validate table: {} for the entity class: {} because keyspace: {} is not present`
so it contains table and entity (not only ks that is `null`). Could you give me an example of how this message could look like to improve it (because I am not sure how to make it better for this case)?

Reply:
> keyspace name present but metadata is out of date
addressed in this commit:  https://github.com/datastax/java-driver/pull/1300/commits/9442728d5e0fa5274c283bf4c1ade0f2ed2c873d
See my comment: https://github.com/datastax/java-driver/pull/1300#discussion_r358248749
for more explanation.

Reply:
Hrmm I don't get it, the two checks are equivalent. From the generated code with the new version:
```java
if(!keyspaceNamePresent(context.getSession().getMetadata().getKeyspaces(), keyspaceId)) {
  ... // warning 1
}
Optional<KeyspaceMetadata> keyspace = context.getSession().getMetadata().getKeyspace(keyspaceId);
if (!keyspace.isPresent()) {
  ... // warning 2
}
```
The two checks to the same thing: if the key is not present in the map, then getKeyspace returns an empty result.

My suggestion was more in the lines of:
```java
CqlIdentifier keyspaceId = context.getKeyspaceId() != null ? context.getKeyspaceId() : context.getSession().getKeyspace().orElse(null);
if (keyspaceId == null) {
  ... // DAO will run unqualified requests on an unqualified session: probably a mapper configuration error, the requests will fail at runtime
}
Optional<KeyspaceMetadata> keyspace = context.getSession().getMetadata().getKeyspace(keyspaceId);
if (!keyspace.isPresent()) {
  ... // what you already have for this case
}
```

Reply:
They are not equivalent. 
1. The first check is checking the case that you mentioned - so in the case of `keyspaceId == null` it will log warn and progress to unqualified requests. But besides that, it is also checking if the keyspace (if provided as not null) is present in the metadata, see:
```
 if(!keyspaceNamePresent(context.getSession().getMetadata().getKeyspaces(), keyspaceId)) {
}
.....
public boolean keyspaceNamePresent(
      Map<CqlIdentifier, KeyspaceMetadata> keyspaces, CqlIdentifier keyspaceId) {
    return keyspaces.keySet().contains(keyspaceId);
  }
```
(`null` will be not present so it will catch your first case.
Please take a look at this test: `SchemaValidationIT#should_log_warning_when_passing_not_existing_keyspace`
2. This second case:
```
 Optional<KeyspaceMetadata> keyspace = context.getSession().getMetadata().getKeyspace(keyspaceId);
    if (!keyspace.isPresent()) {
      LOG.warn("[{}] Unable to validate table: {} for the entity class: {} because keyspace: {} is not present.",
          context.getSession().getName(),
          tableId,
          entityClassName,
          keyspaceId);
      return;
    }
```
is getting the actual `KespaceMetadata` for not null `keyspaceId` (if it was null, then previous case will catch that)

So I think that the current solution is equivalent to your suggestion + it is adding the validation of the existence of the keyspace name.

Reply:
These two conditions are equivalent:
```java
session.getMetadata().getKeyspaces().keySet().contains(keyspaceId)
session.getMetadata().getKeyspace(keyspaceId).isPresent()
```
Except if you assume that:
* the map can contain null values (which is in fact impossible because our implementation relies on Guava's `ImmutableMap`, which doesn't allow it)
* or that there is a bug in `getKeyspace()`.

So the first check already covers the second check. Try to write a test that produces the second warning.

We have two error conditions to cover:
1. the DAO was not built with sufficient information to determine a keyspace id
2. it was, but the keyspace id does not exist

Currently, the first check does both 1 and 2, and the second check does 2 (but is unreachable).

Reply:
Ah, you right - sorry for back and forth on this.
I've switch checks (according to your suggestion here: https://github.com/datastax/java-driver/pull/1300#discussion_r364867139) to first check if `keyspaceId == null` and in that case give a detailed message. Otherwise the `keyspaceNamePresent` check.